### PR TITLE
[BC Break] Refactor code around form submit and multiple buttons

### PR DESCRIPTION
**Pull request by djhi**
We previously had a complex solution for having multiple submit buttons: a SaveContext providing side effects modifiers and also a redirect prop. However, this was redundant and confusing as the `save` function provided by our mutation hooks also accepted side effect override at call time.
Besides, our solution prevented the native browser submit on enter feature and this was an accessibility issue for some users such as Japanese people.

The new solution leverage the fact that we already make the `save` function available through context and users can get the current form values, call the save function directly and override any of the side effects as they usually do with mutation hooks.

We also added side effects props on the SaveButton to make most common cases possible and avoid gotchas such as preventing default event.

- [x] Refactoring
- [x] Upgrade guide
- [x] Documentation
- [x] Update demos
- [x] Rename onFailure to onError for coherence
- [x] Replace onSuccess and onFailure by mutationOptions on SaveButton

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -1411,6 +1411,23 @@ const Layout = (props) => {
...
+
+Besides, our solution prevented the native browser submit on enter feature and this was an accessibility issue for some users such as Japanese people.
+
+The new solution leverage the fact that we already make the `save` function available through context (`useSaveContext`) and users can get the current form values, call the save function directly and override any of the side effects as they usually do with mutation hooks.
+
+We also added side effects props on the SaveButton to make most common cases possible and avoid gotchas such as preventing default event.
+
+If you relied on `handleSubmit` or `handleSubmitWithRedirect`, you can now use the `SaveButton` and override any of the side effect props: `onSuccess`, `onFailure` or `transform`.
```
Comment (fzaninotto) [inline UPGRADE.md:1424]: please add a diff to explicit the changes developers must make in their code

File: packages/ra-core/src/controller/saveContext/usePickSaveContext.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { useMemo } from 'react';
+import pick from 'lodash/pick';
+import { SaveContextValue } from './SaveContext';
+
+export const usePickSaveContext = <
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/saveContext/usePickSaveContext.ts:5]: Could you document the use case for this one? It's not immediate by reading the source

File: packages/ra-core/src/dataProvider/useLoading.ts [hunk 3]
```diff
@@ -17,5 +17,5 @@ import { useIsFetching, useIsMutating } from 'react-query';
 export const useLoading = () => {
     const isFetching = useIsFetching();
     const isMutating = useIsMutating();
-    return isFetching || isMutating;
+    return isFetching > 0 || isMutating > 0;
```
Comment (fzaninotto) [inline packages/ra-core/src/dataProvider/useLoading.ts:20]: Why this change?

File: packages/ra-core/src/form/FormContext.tsx [hunk 4]
```diff
@@ -1,8 +0,0 @@
-import { createContext } from 'react';
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormContext.tsx:1]: this should be documented in Upgrade

File: packages/ra-core/src/form/useGetFormValues.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import { useForm } from 'react-final-form';
+
+/**
+ * This hook returns a function that can be used to get the current values of the form.
+ */
+export const useGetFormValues = <ValuesType = Record<string, any>>() => {
```
Comment (fzaninotto) [inline packages/ra-core/src/form/useGetFormValues.ts:6]: Could we avoid adding this new indirection? I don't want to hide react-final-form, because we're not adding anything better (for now)

File: packages/ra-no-code/src/ResourceConfiguration/ResourceConfiguration.tsx [hunk 6]
```diff
@@ -102,7 +102,7 @@ export const ResourceConfigurationPage = () => {
                 <FormWithRedirect
                     save={save}
                     initialValues={resourceConfiguration}
-                    render={({ handleSubmitWithRedirect }) => (
+                    render={({ handleSubmit }) => (
```
Comment (fzaninotto) [inline packages/ra-no-code/src/ResourceConfiguration/ResourceConfiguration.tsx:105]: this should be documented in the upgrade guide

File: packages/ra-ui-materialui/src/button/SaveButton.spec.tsx [hunk 7]
```diff
@@ -216,161 +155,9 @@ describe('<SaveButton />', () => {
             path: '/customers/123',
             url: '/customers/123',
         },
-        mutationMode: 'pessimistic',
+        mutationMode: 'pessimistic' as MutationMode,
     };
 
-    it('should allow to override the onSuccess side effects', async () => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/button/SaveButton.spec.tsx:222]: You should keep this test

File: packages/ra-ui-materialui/src/button/SaveButton.spec.tsx [hunk 8]
```diff
@@ -165,41 +139,6 @@ describe('<SaveButton />', () => {
         expect(onSubmit).not.toHaveBeenCalled();
     });
 
-    it('should show a notification if the form is not valid', () => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/button/SaveButton.spec.tsx:168]: You should keep this test

File: packages/ra-ui-materialui/src/button/SaveButton.tsx [hunk 9]
```diff
@@ -62,100 +59,56 @@ import { FormRenderProps } from 'react-final-form';
...
     } = props;
-
-    const notify = useNotify();
     const translate = useTranslate();
-    const formContext = useFormContext();
-    const { setOnSuccess, setOnFailure, setTransform } = useSaveContext(props);
-
-    const handleClick = event => {
-        // deprecated: use onSuccess and transform instead of onSave
-        if (typeof onSave === 'function') {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/button/SaveButton.tsx:89]: This should be documented in the upgrade guide

File: packages/ra-ui-materialui/src/button/SaveButton.tsx [hunk 10]
```diff
@@ -62,100 +59,56 @@ import { FormRenderProps } from 'react-final-form';
...
-        } else {
-            if (invalid) {
-                notify('ra.message.invalid_form', { type: 'warning' });
-            }
-            // always submit form explicitly regardless of button type
-            if (event) {
-                event.preventDefault();
-            }
-            handleSubmitWithRedirect(redirect);
+        console.log('defaultPrevented', event.defaultPrevented);
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/button/SaveButton.tsx:85]: console.log() spotted

File: packages/ra-ui-materialui/src/button/SaveButton.tsx [hunk 11]
```diff
@@ -62,100 +59,56 @@ import { FormRenderProps } from 'react-final-form';
...
+        if (event.defaultPrevented) {
+            return;
         }
 
-        if (typeof onClick === 'function') {
-            onClick(event);
+        if (hasSideEffects) {
+            console.log({ hasSideEffects });
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/button/SaveButton.tsx:91]: console.log() spotted

File: packages/ra-ui-materialui/src/button/SaveButton.tsx [hunk 12]
```diff
@@ -62,100 +59,56 @@ import { FormRenderProps } from 'react-final-form';
...
 
-    const type = submitOnEnter ? 'submit' : 'button';
     const displayedLabel = label && translate(label, { _: label });
     return (
         <StyledButton
             className={classnames(SaveButtonClasses.button, className)}
             variant={variant}
             type={type}
-            onClick={handleClick}
-            // TODO: find a way to display the loading state (LoadingButton from mui Lab?)
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/button/SaveButton.tsx:154]: I'd keep this comment: LoadingButton is a nice addition

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: examples/simple/src/comments/CommentCreate.tsx [hunk 13]
```diff
@@ -14,8 +14,8 @@ const now = new Date();
 const defaultSort = { field: 'title', order: 'ASC' };
 
 const CommentCreate = () => (
-    <Create>
-        <SimpleForm redirect={false}>
+    <Create redirect={false}>
```
[the remark this answers is not in the document]
Comment (djhi) [reply examples/simple/src/comments/CommentCreate.tsx:17]: It does

File: packages/ra-core/src/dataProvider/useLoading.ts (hunk 3 above)
Comment (djhi) [reply packages/ra-core/src/dataProvider/useLoading.ts:20]: Because they return numbers where we expect a boolean

File: packages/ra-ui-materialui/src/button/SaveButton.tsx (hunk 12 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/button/SaveButton.tsx:154]: https://mui.com/components/buttons/#loading-button

File: packages/ra-ui-materialui/src/button/SaveButton.spec.tsx (hunk 8 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/button/SaveButton.spec.tsx:168]: Nope, it does not show any notification anymore, we entirely rely on the form validation mechanisms

File: packages/ra-ui-materialui/src/button/SaveButton.spec.tsx (hunk 7 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/button/SaveButton.spec.tsx:222]: Added back

File: docs/CreateEdit.md [hunk 14]
```diff
@@ -1674,8 +1674,6 @@ export const PostEdit = () => (
 Here are the props received by the `Toolbar` component when passed as the `toolbar` prop of the `SimpleForm` or `TabbedForm` components:
 
 * `alwaysEnableSaveButton`: Force enabling the `<SaveButton>`. If it's not defined, the `<SaveButton>` will be enabled using the `pristine` prop (disabled if pristine, enabled otherwise).
-* `handleSubmitWithRedirect`: The function to call in order to submit the form. It accepts a single parameter overriding the form's default redirect.
-* `handleSubmit` which is the same prop as in [react-final-form](https://final-form.org/docs/react-final-form/types/FormRenderProps#handlesubmit)
 * `invalid`: A boolean indicating whether the form is invalid
 * `pristine`: A boolean indicating whether the form is pristine (eg: no inputs have been changed yet)
 * `redirect`: The default form's redirect
```
Comment (fzaninotto) [inline docs/CreateEdit.md:1679]: Thios one should be moved to the Create and Edit components

File: docs/CreateEdit.md [hunk 15]
```diff
@@ -2282,51 +2274,6 @@ const dataProvider = {
 }
 ```
 
-### Using `onSave` To Alter the Form Submission Behavior
```
Comment (fzaninotto) [inline docs/CreateEdit.md:2285]: The proper way to do this must be documented somewhere

File: examples/demo/src/orders/OrderEdit.tsx [hunk 16]
```diff
@@ -84,143 +84,152 @@ const OrderForm = (props: any) => {
...
-                                    </Typography>
-                                    <Grid container>
-                                        <Grid item xs={12} sm={12} md={6}>
-                                            <Labeled
-                                                source="date"
-                                                resource="commands"
-                                            >
-                                                <DateField
+            render={({ handleSubmit, ...formProps }: any) => (
+                <form onSubmit={handleSubmit}>
```
Comment (fzaninotto) [inline examples/demo/src/orders/OrderEdit.tsx:88]: That's a breaking change, too, and a big one. I thought SimpleForm and others included a `<form>` tag. Can you revert this BC break?

File: examples/demo/src/reviews/ReviewEditToolbar.tsx [hunk 17]
```diff
@@ -43,10 +42,11 @@ const ReviewEditToolbar = (props: ToolbarProps<Review>) => {
             ) : (
                 <Fragment>
                     <SaveButton
-                        handleSubmitWithRedirect={handleSubmitWithRedirect}
                         invalid={invalid}
                         saving={saving}
-                        redirect="list"
+                        onSuccess={() => {
+                            redirect('list', '/reviews');
```
Comment (fzaninotto) [inline examples/demo/src/reviews/ReviewEditToolbar.tsx:48]: you're missing the default notification

File: UPGRADE.md [hunk 18]
```diff
@@ -1411,6 +1411,71 @@ const Layout = (props) => {
...
+```diff
+const PostCreateToolbar = props => {
++    const redirect = useRedirect();    
+    return (
+        <Toolbar {...props}>
+            <SaveButton
+                label="post.action.save_and_edit"
+-                redirect="edit"
++                onSuccess={data => {
++                    redirect('edit', '/posts', data.id)
```
Comment (fzaninotto) [inline UPGRADE.md:1454]: this is missing the default notification

File: packages/ra-ui-materialui/src/button/SaveButton.spec.tsx (hunk 8 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/button/SaveButton.spec.tsx:168]: Then it should be moved to Create & Edit tests

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: packages/ra-core/src/controller/create/CreateBase.spec.tsx [hunk 19]
```diff
@@ -22,215 +20,163 @@ describe('CreateBase', () => {
...
+            ),
         });
         const onSuccess = jest.fn();
         const onSuccessOverride = jest.fn();
 
-        const SetOnSuccess = () => {
-            const saveContext = useSaveContext();
-
-            useEffect(() => {
-                saveContext.setOnSuccess(onSuccessOverride);
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-core/src/controller/create/CreateBase.spec.tsx:65]: As you can provide default side effects at the `Create`, `Edit` and `SaveButton` level, I fail to find a use case for this

File: packages/ra-core/src/controller/create/CreateBase.spec.tsx [hunk 20]
```diff
@@ -22,215 +20,163 @@ describe('CreateBase', () => {
...
+
+        await waitFor(() => {
+            expect(dataProvider.create).toHaveBeenCalledWith('posts', {
+                data: { test: 'test' },
+            });
+        });
     });
 
-    it('should allow to override the onSuccess function', () => {
+    it('should allow to override the onSuccess function', async () => {
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-core/src/controller/create/CreateBase.spec.tsx:54]: Done

File: packages/ra-core/src/controller/edit/EditBase.spec.tsx [hunk 21]
```diff
@@ -24,252 +20,230 @@ describe('EditBase', () => {
...
+        });
     });
 
-    it('should allow to override the onSuccess function', () => {
-        const dataProvider = ({
-            getOne: () => Promise.resolve({ data: { id: 12 } }),
-            update: (_, { id, data, previousData }) =>
-                Promise.resolve({ data: { id, ...previousData, ...data } }),
-        } as unknown) as DataProvider;
+    it('should allow to override the onSuccess function', async () => {
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-core/src/controller/edit/EditBase.spec.tsx:70]: Done

File: packages/ra-no-code/src/ResourceConfiguration/ResourceConfiguration.tsx (hunk 6 above)
Comment (djhi) [reply packages/ra-no-code/src/ResourceConfiguration/ResourceConfiguration.tsx:105]: You mean the difference between `handleSubmit` and `handleSubmitWithRedirect` and the fact that `handleSubmitWithRedirect` was removed? If yes, it already is

File: packages/ra-ui-materialui/src/button/SaveButton.spec.tsx (hunk 8 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/button/SaveButton.spec.tsx:168]: As discussed, I moved the notification in the FormWithRedirect component

File: docs/CreateEdit.md (hunk 15 above)
Comment (djhi) [reply docs/CreateEdit.md:2285]: We already have documentation for `transform` right ?

File: examples/demo/src/orders/OrderEdit.tsx (hunk 16 above)
Comment (djhi) [reply examples/demo/src/orders/OrderEdit.tsx:88]: `SimpleForm` and `TabbedForm` includes the `form` tag but here, we're using `FormWithRedirect`. No we can't remove this BC break as it is required to ensure native submit on enter for instance

File: packages/ra-core/src/controller/saveContext/usePickSaveContext.ts (hunk 2 above)
Comment (djhi) [reply packages/ra-core/src/controller/saveContext/usePickSaveContext.ts:8]: Done but all or other `usePickXXX` are undocumented

File: packages/ra-core/src/form/FormContext.tsx (hunk 4 above)
Comment (djhi) [reply packages/ra-core/src/form/FormContext.tsx:1]: Done

File: packages/ra-core/src/form/useGetFormValues.ts (hunk 5 above)
Comment (djhi) [reply packages/ra-core/src/form/useGetFormValues.ts]: Removed but I feel bad about it

File: packages/ra-ui-materialui/src/button/SaveButton.tsx (hunk 9 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/button/SaveButton.tsx:89]: Done

File: examples/demo/src/reviews/ReviewEditToolbar.tsx (hunk 17 above)
Comment (djhi) [reply examples/demo/src/reviews/ReviewEditToolbar.tsx:57]: Added

File: UPGRADE.md (hunk 18 above)
Comment (djhi) [reply UPGRADE.md:1507]: Added

File: UPGRADE.md [hunk 22]
```diff
@@ -1411,6 +1431,124 @@ const Layout = (props) => {
...
+### `<FormContext>`, `<FormContextProvider>` and `useFormContext` Have Been Removed
+
+As the `save` and `saving` properties were already available through the `<SaveContext>` component and its `useSaveContext` hook, we removed the `<FormContext>`, `<FormContextProvider>` components as well the `useFormContext` hook.
+
+The functions around form groups management we extracted into the `<FormGroupsProvider>` component and `useFormGroups` hook.
+
+### The `redirect` prop Has Been Removed From `FormWithRedirect`, `SimpleForm`, `TabbedForm` and `SaveButton`
+
+The `FormWithRedirect`, `SimpleForm`, `TabbedForm` and `SaveButton` don't have a `redirect` prop anymore. You should handle redirection either:
+- in the submit handler if you provided a custom one, leveraging the `useRedirect` hook
```
Comment (fzaninotto) [inline UPGRADE.md:1494]: As you explicited 3 of the 3 list items with examples below, why don't you do the same fpor the 3rd list item, and femove the list altogether? It'sll be easier to follow. 

Otherwise, for each list item, I think 'how am I going to do this?'

File: UPGRADE.md [hunk 23]
```diff
@@ -1411,6 +1431,124 @@ const Layout = (props) => {
...
++                    redirect('edit', '/posts', data.id)
++                }}
+            />
+        </Toolbar>
+    );
+};
+```
+
+### The Form Components `save` Prop Has Been Renamed to `onSubmit`:
+
```
Comment (fzaninotto) [inline UPGRADE.md:1536]: You should probably add that this prop was only useful if SimpleForml was used in standalone, not as child of Create of Edit. In the general case, there is nothing to do.

**Review by fzaninotto — CHANGES_REQUESTED** [review]
We need to provide more guidance, and to be reassuring, in the upgrade guide.

File: docs/CreateEdit.md [hunk 24]
```diff
@@ -765,7 +766,7 @@ This custom Edit view has no action buttons or aside component - it's up to you
 
 ## The `<SimpleForm>` component
 
-The `<SimpleForm>` component receives the `record` as prop from its parent component. It is responsible for rendering the actual form. It is also responsible for validating the form data. Finally, it receives a `handleSubmit` function as prop, to be called with the updated record as an argument when the user submits the form.
```
Comment (fzaninotto) [inline docs/CreateEdit.md:768]: > The `<SimpleForm>` component receives the `record` as prop from its parent component

It actually reads it from the RecordContext

File: docs/CreateEdit.md (hunk 15 above)
Comment (fzaninotto) [reply docs/CreateEdit.md:2285]: right

File: UPGRADE.md [hunk 25]
```diff
@@ -1411,6 +1431,124 @@ const Layout = (props) => {
...
+We previously had a complex solution for having multiple submit buttons: a `SaveContext` providing side effects modifiers and refs to the current ones. However, this was redundant and confusing as the `save` function provided by our mutation hooks also accept side effect override at call time.
+
+We also supported a redirect prop both on the form component and on the `<SaveButton>`. It was even more confusing when adding custom or multiple `<SaveButton>` as those received two submission related functions: `handleSubmit` and `handleSubmitWithRedirect`.
+
+Besides, our solution prevented the native browser submit on enter feature and this was an accessibility issue for some users such as Japanese people.
+
+The new solution leverage the fact that we already have the `save` function available through context (`useSaveContext`). You can get the current form values using [final-form hooks](https://final-form.org/docs/react-final-form/api/useFormState), call the `save` function directly and override any of the side effects as you usually do with mutation hooks.
+
+### `<SaveButton>` Props and Behavior Have Changed
```
[the remark this answers is not in the document]
Comment (djhi) [reply UPGRADE.md:1444]: I disagree, there are more changes than those two props. I actually forgot about `mutationOptions`

[the remark this answers is not in the document]
Comment (djhi) [reply UPGRADE.md:1444]: I'll make two sections

File: packages/ra-core/src/controller/edit/useEditController.ts [hunk 26]
```diff
@@ -7,7 +7,8 @@ import {
     Record,
     MutationMode,
     OnSuccess,
-    OnFailure,
+    onError,
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/edit/useEditController.ts:10]: A type starting with a lowercase letter? Feels incorrect.

File: packages/ra-ui-materialui/src/button/SaveButton.tsx [hunk 27]
```diff
@@ -34,19 +32,11 @@ import { FormRenderProps } from 'react-final-form';
  * @prop {string|boolean} redirect Override of the default redirect in case of success. Can be 'list', 'show', 'edit' (for create views), or false (to stay on the creation form).
  * @prop {function} onSave (deprecated)
  * @prop {function} onSuccess Callback to execute instead of the default success side effects. Receives the dataProvider response as argument.
- * @prop {function} onFailure Callback to execute instead of the default error side effects. Receives the dataProvider error response as argument.
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/button/SaveButton.tsx:37]: this is incorrect: SaveButton now only receives mutationOptions

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: UPGRADE.md [hunk 28]
```diff
@@ -1411,6 +1431,124 @@ const Layout = (props) => {
...
+    const createControllerProps = useCreateController();
+    const notify = useNotify();
++    const redirect = useRedirect();
+
+    const handleSubmit = (values) => {
+-        createControllerProps.save(values, 'show', {
++        createControllerProps.save(values, {
+            onSuccess: (data) => {
+                notify('Success');
++                redirect('show', '/posts', data.id);
```
[the remark this answers is not in the document]
Comment (djhi) [reply UPGRADE.md:1465]: Because the previous code of the diff had the `redirect` prop

File: packages/ra-ui-materialui/src/button/SaveButton.tsx (hunk 27 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/button/SaveButton.tsx:37]: Both onFailure and onSuccess have been replaced with mutationOptions

File: packages/ra-core/src/controller/edit/useEditController.ts (hunk 26 above)
Comment (djhi) [reply packages/ra-core/src/controller/edit/useEditController.ts:10]: And should be removed anyway

File: UPGRADE.md (hunk 23 above)
Comment (djhi) [reply UPGRADE.md:1655]: Added

**Comment by fzaninotto** [comment]
Conflict on CreateEdit.md
